api/unstable/fingerprint.h

```diff
@@ -74,3 +74,28 @@ int s2n_client_hello_get_fingerprint_hash(struct s2n_client_hello *ch,
 int s2n_client_hello_get_fingerprint_string(struct s2n_client_hello *ch,
         s2n_fingerprint_type type, uint32_t max_size,
         uint8_t *output, uint32_t *output_size);
+
+/**
+ * Creates an s2n_client_hello from bytes representing a ClientHello message.
+ *
+ * Unlike s2n_connection_get_client_hello, the s2n_client_hello produced by this
+ * method is owned by the application and must be freed with s2n_client_hello_free.
+ *
+ * This method does not support SSLv2 ClientHellos.
+ *
+ * @param bytes The raw bytes representing the ClientHello.
+ * @param size The size of raw_message.
+ * @returns A new s2n_client_hello on success, or NULL on failure.
+ */
+struct s2n_client_hello *s2n_client_hello_parse_bytes(uint8_t *bytes, uint32_t size);
```

Comment:
If possible, it would be helpful if this function(or another) could guide the caller on how many more bytes are needed to parse the client hello.

I know this can be solved with TLS record parsing in the application, but even that kind of code would ideally live inside of s2n and the application can just supply bytes.

Reply:
At the moment, the use case we have is that customers have complete ClientHello messages, not unparsed records. But I realized the first version didn't actually expect complete ClientHello messages-- I forgot about the message header!

Parsing the client hello from TLS records is an interesting use case, but I think it's out of scope for now. However, in case we want to handle it in the future, I've renamed the function to make it clear that it operates on messages, not records. You're right that records would require a different signature with some way to indicate "partial success".